File: mvp-java/ui/src/main/java/com/wasisto/githubuserfinder/android/ServiceLocator.java
package com.wasisto.githubuserfinder.android;

import androidx.lifecycle.ViewModelProvider;

import com.wasisto.githubuserfinder.common.Logger;
import com.wasisto.githubuserfinder.domain.interfaces.GithubDataSource;
import com.wasisto.githubuserfinder.domain.interfaces.SearchHistoryDataSource;

import java.util.concurrent.ExecutorService;

public interface ServiceLocator {

    GithubDataSource getGithubDataSource();

    SearchHistoryDataSource getSearchHistoryDataSource();

    ExecutorService getIoExecutorService();

    ExecutorService getMainExecutorService();

    Logger getLogger(Class<?> cls);
}

File: mvvm-kotlin/ui/src/main/java/com/wasisto/githubuserfinder/android/ServiceLocator.kt
package com.wasisto.githubuserfinder.android

import com.wasisto.githubuserfinder.common.Logger
import com.wasisto.githubuserfinder.domain.interfaces.GithubDataSource
import com.wasisto.githubuserfinder.domain.interfaces.SearchHistoryDataSource
import kotlinx.coroutines.CoroutineDispatcher
import kotlin.reflect.KClass

interface ServiceLocator {

    val githubDataSource: GithubDataSource

    val searchHistoryDataSource: SearchHistoryDataSource

    val ioDispatcher: CoroutineDispatcher

    val mainDispatcher: CoroutineDispatcher

    fun getLogger(cls: KClass<*>): Logger
}